### ISSUE TITLE
lmod and path lengths advice

### ISSUE THREAD

**Issue by svenevs** [spack/spack#6093]
I'm curious as to how other people are actually using `lmod` with `spack`.  Initially I thought to myself "hey since I've built all this cool stuff and it's tuned to this rig, may as well use it!"  I made an "uber-module" that basically loads in everything possible.  Doing so results in errors such as `man man` -> `MANPATH too long` or worse, silent errors on `PKG_CONFIG_PATH` because the entry shows up at the end.

I had initially just ignored it since I can live without man pages, or purge if I really need one.  Today I finally realized that the `PKG_CONFIG_PATH` thing is a bigger problem, but ultimately caused by the same thing (too many paths getting prepended by loading in all of these modules).

Clearly loading every single package always may not really be smart.  I'm just curious if there is any high level thoughts on what should be loaded vs what is not necessary.  For example, if I build `opencv+python`, then I'll want to make sure that I load both `opencv` and the `python` that was used to build it.  Eigen as well, and a few others like the image libraries, sure.  But e.g. a full blown `spack module loads -m lmod -r opencv %gcc` brings in way more than I actually need, meaning that if I load all of the dependencies of the core packages I *actually* want, I'll exceed the linux path lengths.

Any wisdom on how to select which ones should / should not be loaded?

**Comment by alalazo** [comment]
Would this be solved if we implement #8815 ? If we generate module files that load their run dependencies by default you could just `spack load opencv` and `python` will be loaded as a consequence.

**Comment by healther** [comment]
I'm not sure that this will be helping, because internally the modules will still append the respective environment variables. Up to a point this is fundamental to the way modules work. We ran into similar problems in the sense that autocompletion took forever because the system was going through all (several hundred) directories on `PATH` on a rather slow nfs filesystem. 

The solution was to implement the `spack view` command and manually write a module for that view. Essentially this links all packages (and their dependencies) into one folder that then looks similar to `/usr`. Ideally this would be integrated in the environments concept and then the user facing side of `spack` would look a lot like virtualenv (with an `spack env activate myfavoriteenvironment` command being all that is required). I'm uncertain what the latest status on this is though.

**Comment by svenevs** [comment]
FWIW the resolution on my end was to be more careful about what modules are automatically loaded at startup.  I also had significant problems with `$PYTHONPATH` using modules + python with spack, and as such `activate` all of my python packages.  My personal workflow is to use spack to install the core computational libraries for python that I need (such as `numpy` and `scipy` etc), and I use `py-pip` to install everything else I need globally (or more often, just use virtual environments).  Just noting for the records that python installs using modules are much more likely to max out the path length for any reasonably sized python stack.  But to be extra clear, I can get away with activating or pip installing things because I'm the only person using this spack instance.  Larger sites will have to be more careful.

In short: this is not a spack problem, users need to be judicious in which modules they are loading to avoid path length limitations.

**Comment by alalazo** [comment]
> My personal workflow is to use spack to install the core computational libraries for python that I need (such as numpy and scipy etc), and I use py-pip to install everything else I need globally (or more often, just use virtual environments).

For what is worth that's also what we do on our clusters.
